providers/implementations/signature/eddsa_sig.c

```diff
@@ -273,11 +429,20 @@ static int ed25519_digest_sign(void *vpeddsactx, unsigned char *sigret,
 #endif /* S390X_EC_ASM */
 
     if (peddsactx->prehash_flag) {
-        if (!EVP_Q_digest(peddsactx->libctx, SN_sha512, NULL, tbs, tbslen, md, &mdlen)
+        if (!peddsactx->prehash_by_caller_flag) {
+            if (!EVP_Q_digest(peddsactx->libctx, SN_sha512, NULL, tbs, tbslen, md, &mdlen)
                 || mdlen != EDDSA_PREHASH_OUTPUT_LEN)
+                return 0;
+            tbs = md;
+            tbslen = mdlen;
+        } else if (tbslen != EDDSA_PREHASH_OUTPUT_LEN) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST_LENGTH);
             return 0;
-        tbs = md;
-        tbslen = mdlen;
+        }
+    } else if (peddsactx->prehash_by_caller_flag) {
+        /* The caller is supposed to set up a ph instance! */
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_INSTANCE);
```

Comment:
I rather think that the error message for this situation should be much more detailed. `invalid instance` does not convey any context. Closer might be `Invalid EdDSA instance for attempted operation` or something more apt...

Reply:
Sure